### ISSUE TITLE
Document the workflow for retargeting

### ISSUE THREAD

**Issue by fire** [DeepMotionEditing/deep-motion-editing#88]
Hi, I tried the demo's and the demo shows as described the IK re-targeting.

However, I wanted to understand how to make a new target skeleton something we can move animations to.

There's many pieces of information scattered in the issue tracker and the re-targeting process is hard to understand.

I'm trying to writing a guide on how to use the retargetter.

Workflow:

Inputs:
* A trained deep motion editing ML model
* A specific VRM skeleton animation that has been preprocessed to strip non-essential information
* Animation of a specific VRM skeleton model
    * The process of using another tool to retarget the VRM skeleton to train retargeting is complicated because it requires a CUDA install and a separate re-targeting tool. Can this step be removed?
* Following the instructions to add the four steps to the bvh importer.

Goal:

Input the trained model and give animations on any of the trained rigs and have it be on the new rig.

Other:

How many animations is necessary to train a new skeleton?

What is the expected time to retrain?

Also is it possible to standardize on gltf2 for the skeleton information? 

I also have problems with the bvh being in the wrong orientation with respect to the floor and with cm to meter scaling.

Is it possible to store this as an ONNX model for game runtime retargeting?

**Comment by PeizhuoLi** [comment]
First of all, thanks for your interest in this project.

> The process of using another tool to retarget the VRM skeleton to train retargeting is complicated because it requires a CUDA install and a separate re-targeting tool. Can this step be removed?

The purpose for this part is mainly converting the fbx file, which is binary, not easy for python to parse and might contains skin information that is not used in the retargeting process, to bvh file, which is plain text, contains only the skeleton and animation information. In the first step we use Blender, an outside software. Then the further script collect the information and save it in a compact numpy array to make it easier to access by training. I'm sure there is an easy approach to bypass CUDA installation.

> How many animations is necessary to train a new skeleton?

In our experiments, we successfully trained a model on a dataset with around 9000 frames (30fps). If the data is diverse enough ideally this amount will be sufficient to converge.

> Also is it possible to standardize on gltf2 for the skeleton information?

We use bvh file, which is commonly used by animation community and can be parsed by mainstream 3D softwares like Blender and Maya. Although I'm not familiar with gltf2 that you mentioned, I believe it's possible.

> I also have problems with the bvh being in the wrong orientation with respect to the floor and with cm to meter scaling.

The orientation can be problematic. I'd suggest you to manually correct it before retargeting. However, since we used per-character normalization, I don't think cm to meter scaling will be an obstacle.

> Is it possible to store this as an ONNX model for game runtime retargeting?

In this project we don't target a realtime setup. We are using 1D temporal convolution, meaning to generate a specific frame the network will look at a few frames before and following the the current frame (~15 frames in our implementation). If you are trying to retarget motion with low latency, current implementation might not be feasible. It would be interesting to try to replace `casual convolution` with current convolution, but it's beyond the scope of this project.


Again, thanks for your questions and sorry for our delayed reply :)

**Comment by fire** [comment]
The process of look at a few frames before and following doesn't seem interactive, but using it to process animation as a batch tool would be usable. Like in an editor record frames and then wait < 1 minute for the results.

**Comment by PeizhuoLi** [comment]
I'm glad to see that this project will become a useful plugin! However, I don't know ONNX or TorchScrip well so it's not clear to me why the requires_grad causes the trouble. But in my mind, if you only try to remove python dependency in test time, requires_grad doesn't make a difference cause it only affects training time. And if it's really problematic, you can provide more details on the variables and it's possible that I can offer some advice.

**Comment by fire** [comment]
I have another question.

You mentioned unpaired animations. Does this mean I can have set of rigs with > 9000 frames of animations and the training will work?

**Comment by PeizhuoLi** [comment]
Ideally it should be enough: in one of our demos we trained a model with a dataset of approximate 10,000 frames and the result looks good. 

Regarding no corresponding, it's not a problem. Our model is designed to work with unpaired motion retargeting.

**Comment by fire** [comment]
Any approaches for implementing multiple skeletons?

> Regarding the pre-trained model settings, you are right. It was trained on skeleton_type=0 and skeleton_type=10.

> I totally agree that changing the naming rules will be much clean. But in the cleaning work that we are doing now, we add an extra parameter in the eval_single_pair.py for user to manually specify the group of given character instead of using the character name trick, mainly considering in test time a network can only support two types of skeleton.

The goal is to have skeleton type (which is an enum.. I would prefer string.) for each side of the pair so I can have more than two skeletons in the training.

**Comment by fire** [comment]
There is a limitation where DeepMotionEditing appears to want paired animation bytes when demoing.

Not sure how to resolve.

**Comment by fire** [comment]
Resolved the problems. I was able to have some results, but training the machine learning model takes time.
